File: ptn/tps.go
package ptn

import (
	"errors"
	"fmt"
	"strconv"
	"strings"

	"github.com/nelhage/taktician/tak"
)

func ParseTPS(tpn string) (*tak.Position, error) {
	var pieces [][]tak.Square
	words := strings.Split(tpn, " ")
	if len(words) != 3 {
		return nil, errors.New("bad TPN: wrong number of words")
	}
	turn, err := strconv.Atoi(words[1])
	if err != nil {
		return nil, fmt.Errorf("bad turn: %s", words[1])
	}
	if turn != 1 && turn != 2 {
		return nil, fmt.Errorf("bad turn: %s", words[1])
	}
	move, err := strconv.Atoi(words[2])
	if err != nil {
		return nil, fmt.Errorf("bad move: %s", words[2])
	}
	move = 2*(move-1) + (turn - 1)

	rows := strings.Split(words[0], "/")
	for _, r := range rows {
		row, err := parseRow(r)
		if err != nil {
			return nil, err
		}
		pieces = append([][]tak.Square{row}, pieces...)
	}
	if len(pieces) < 3 || len(pieces) > 8 {
		return nil, fmt.Errorf("bad size board: %d", len(pieces))
	}
	for i, r := range pieces {
		if len(r) != len(pieces) {
			return nil, fmt.Errorf("row %d bad length: %d", i, len(r))
		}
	}
	return tak.FromSquares(tak.Config{Size: len(pieces)}, pieces, move)
}

func FormatTPS(p *tak.Position) string {
	var rows []string
	for i := p.Size() - 1; i >= 0; i-- {
		rows = append(rows, tpsRow(p, i))
	}
	var toMove string
	if p.ToMove() == tak.White {
		toMove = "1"
	} else {
		toMove = "2"
	}
	return fmt.Sprintf("%s %s %d", strings.Join(rows, "/"), toMove, p.MoveNumber()/2+1)
}

func tpsRow(p *tak.Position, y int) string {
	var bits []string
	for x := 0; x < p.Size(); {
		var i int
		for i = 0; x+i < p.Size() && len(p.At(x+i, y)) == 0; i++ {
		}
		switch i {
		case 0:
			bits = append(bits, tpsSquare(p.At(x, y)))
			x++
		case 1:
			bits = append(bits, "x")
		default:
			bits = append(bits, fmt.Sprintf("x%d", i))
		}
		x += i
	}
	return strings.Join(bits, ",")
}

func tpsSquare(sq tak.Square) string {
	var out []byte
	for i := len(sq) - 1; i >= 0; i-- {
		if sq[i].Color() == tak.White {
			out = append(out, '1')
		} else {
			out = append(out, '2')
		}
	}
	if sq[0].Kind() == tak.Standing {
		out = append(out, 'S')
	} else if sq[0].Kind() == tak.Capstone {
		out = append(out, 'C')
	}
	return string(out)
}

func parseRow(row string) ([]tak.Square, error) {
	var out []tak.Square
	bits := strings.Split(row, ",")
	for _, bit := range bits {
		if bit[0] == 'x' {
			count := 1
			if len(bit) > 1 {
				count = int(bit[1] - '0')
			}
			for i := 0; i < count; i++ {
				out = append(out, nil)
			}
			continue
		}
		stack := make(tak.Square, len(bit))
		for i, b := range bit {
			switch b {
			case '1':
				stack[len(stack)-i-1] = tak.MakePiece(tak.White, tak.Flat)
			case '2':
				stack[len(stack)-i-1] = tak.MakePiece(tak.Black, tak.Flat)
			case 'C', 'S':
				if i != len(bit)-1 {
					return nil, fmt.Errorf("stone type not at end of stack: %s", bit)
				}
				stack = stack[1:]
				color := stack[0].Color()
				if b == 'S' {
					stack[0] = tak.MakePiece(color, tak.Standing)
				} else {
					stack[0] = tak.MakePiece(color, tak.Capstone)
				}
			default:
				return nil, fmt.Errorf("malformed stack: %s", bit)
			}
		}
		out = append(out, stack)
	}
	return out, nil
}

File: python/tak/ptn/tps.py
import tak


def parse_tps(tps):
    bits = tps.split(" ")
    if len(bits) != 3:
        raise IllegalTPS("need three components")
    board, who, move = bits

    if not who in "12":
        raise IllegalTPS("Current player must be either 1 or 2")
    try:
        ply = 2 * (int(move) - 1) + int(who) - 1
    except ValueError:
        raise IllegalTPS("Bad move number: " + move)

    squares = []
    rows = board.split("/")
    for row in reversed(rows):
        rsq = parse_row(row)
        if len(rsq) != len(rows):
            raise IllegalTPS("inconsistent size")
        squares += rsq

    return tak.Position.from_squares(tak.Config(size=len(rows)), squares, ply)


def parse_row(rtext):
    squares = []
    bits = rtext.split(",")
    for b in bits:
        if b[0] == "x":
            n = 1
            if len(b) > 1:
                n = int(b[1:])
            squares += [[]] * n
            continue

        stack = []
        for c in b:
            if c == "1":
                stack.append(tak.Piece.cached(tak.Color.WHITE, tak.Kind.FLAT))
            elif c == "2":
                stack.append(tak.Piece.cached(tak.Color.BLACK, tak.Kind.FLAT))
            elif c in ("C", "S"):
                if not stack:
                    raise IllegalTPS("bare capstone or standing")
                typ = tak.Kind.CAPSTONE if c == "C" else tak.Kind.STANDING
                stack[-1] = tak.Piece.cached(stack[-1].color, typ)
            else:
                raise IllegalTPS("bad character: " + c)

        squares.append(list(reversed(stack)))
    return squares


def format_tps(pos):
    rows = []
    for row in range(pos.size):
        i = row * pos.size
        rows.append(_format_row(pos.board[i : i + pos.size]))

    return " ".join(
        ["/".join(reversed(rows)), str((pos.ply % 2) + 1), str(pos.ply // 2 + 1)]
    )


def _format_row(row):
    out = []
    i = 0
    while i < len(row):
        x = 0
        while i + x < len(row) and row[i + x] == []:
            x += 1
        if x > 0:
            out.append("x{0}".format(x if x > 1 else ""))
            i += x
        else:
            out.append(_format_square(row[i]))
            i += 1
    return ",".join(out)


def _format_square(sq):
    out = []
    for p in reversed(sq):
        if p.color == tak.Color.WHITE:
            out.append("1")
        else:
            out.append("2")

    if sq[0].kind == tak.Kind.STANDING:
        out.append("S")
    elif sq[0].kind == tak.Kind.CAPSTONE:
        out.append("C")

    return "".join(out)


class IllegalTPS(Exception):
    pass


__all__ = ["parse_tps", "format_tps", "IllegalTPS"]
